src/crimson/osd/ops_executer.cc

```diff
@@ -740,96 +741,128 @@ version_t OpsExecuter::get_last_user_version() const
   return pg->get_last_user_version();
 }
 
-void OpsExecuter::make_writeable(std::vector<pg_log_entry_t>& log_entries)
+// a counterpart of the `make_writeable()` in the classical OSD.
+// the main difference is this one executes before, not after,
+// all other operations. the rationale behind the departure is
+// simplicity (no new / old obs, no txn reordering).
+//
+// TODO: make this `static`
+std::unique_ptr<OpsExecuter::CloningContext> OpsExecuter::execute_clone(
+  const SnapContext& snapc,
+  const ObjectState& initial_obs,
+  const SnapSet& initial_snapset,
+  PGBackend& backend,
+  ceph::os::Transaction& txn)
 {
-  const hobject_t& soid = obc->obs.oi.soid;
+  std::unique_ptr<CloningContext> cloning_ctx;
+  const hobject_t& soid = initial_obs.oi.soid;
   logger().debug("{} {} snapset={} snapc={}",
                  __func__, soid,
-                 obc->ssc->snapset, snapc);
-
-  // clone?
-  if (head_os.exists &&                          // old obs.exists
-      snapc.snaps.size() &&                      // there are snaps
-      snapc.snaps[0] > obc->ssc->snapset.seq) {  // existing obj is old
-
-    // clone object, the snap field is set to the seq of the SnapContext
-    // at its creation.
-    hobject_t coid = soid;
-    coid.snap = snapc.seq;
-
-    // existing snaps are stored in descending order in snapc,
-    // cloned_snaps vector will hold all the snaps stored until snapset.seq
-    const std::vector<snapid_t> cloned_snaps = [&] {
-      auto last = std::find_if(
-        std::begin(snapc.snaps), std::end(snapc.snaps),
-        [&](snapid_t snap_id) { return snap_id <= obc->ssc->snapset.seq; });
-      return std::vector<snapid_t>{std::begin(snapc.snaps), last};
-    }();
-
-    // version
-    osd_op_params->at_version = pg->next_version();
+                 initial_snapset, snapc);
+
+  cloning_ctx = std::make_unique<CloningContext>();
```

Comment:
i'd suggest drop the line of `std::unique_ptr<CloningContext> cloning_ctx;` to avoid the (smallish) overhead of 

1. the assignment operator 
2. the creation of a temporary object (`std::make_unique<CloningContext>`)
3. the destruction of a temporary object (`std::make_unique<CloningContext>`)

in other words, just put
```c++
auto cloning_ctx = std::make_unique<CloningContext>();
```

Reply:
Fixed. Thanks!